direct/core/src/main/java/cz/o2/proxima/direct/core/BulkAttributeWriter.java

```diff
@@ -59,4 +71,8 @@ default Type getType() {
    * @param watermark timestamp of the new watermark
    */
   default void updateWatermark(long watermark) {}
+
+  @SuppressWarnings("unchecked")
+  @Override
+  Factory asFactory(RepositoryFactory repositoryFactory);
```

Comment:
I'm not big fan of `RepositoryFactory` being serialized in all reader factories. Why don't we have `Factory#create(Repository)` instead?

I think we should be aiming for following:
- RepositoryFactory is only used in top level classes where there is no other way of retrieving repository (eg. source)
- Lower level factories (readers, ...), should get materialized repository instance, to keep the UDF size low.

I think this may also lead to a cleaner design.

WDYT?

Reply:
Makes sense. Will try that.

Reply:
I see the issue with that approach. There are cases in which you need to deserialize the reader/writer without reference to Repository. One example of that is `DirectAttributeFamilyDescriptor`. I can include RepositoryFactory in the descriptor, but that would mean change in `serialVersionUID`, because otherwise it would cause NPEs. But we maybe might change it now, because this change is already breaking to serialization. Will try the to go down the path to see if I find any more issues.

Reply:
I wouldn't worry about that, because this is a huge breaking change anyway (that should help to avoid future BCs). 👍